extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -559,18 +559,15 @@ private void discoverTasks()
         KafkaIndexTask kafkaTask = (KafkaIndexTask) task;
         String taskId = task.getId();
 
-        // determine which task group this task belongs to and do a consistency check on partitions
-        Integer taskGroupId = null;
-        for (Integer partition : kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap().keySet()) {
-          if (taskGroupId == null) {
-            taskGroupId = getTaskGroupIdForPartition(partition);
-          } else if (!taskGroupId.equals(getTaskGroupIdForPartition(partition))) {
-            log.warn("Stopping task [%s] which does not match the expected partition allocation", taskId);
-            stopTask(taskId, false);
-            taskGroupId = null; // prevents the next block of code from adding the item to taskGroups
-            break;
-          }
-        }
+        // Determine which task group this task belongs to based on one of the partitions handled by this task. If we
+        // later determine that this task is actively reading, we will make sure that it matches our current partition
+        // allocation (getTaskGroupIdForPartition(partition) should return the same value for every partition being read
+        // by this task) and kill it if it is not compatible. If the task is instead found to be in the publishing
+        // state, we will permit it to complete even if it doesn't match our current partition allocation to support
+        // seamless schema migration.
+
+        Iterator<Integer> it = kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap().keySet().iterator();
```

Comment:
The main change in this PR is to move this consistency check so that it is only applied to reading tasks and not publishing tasks. Without this change, if the user posts an updated configuration that changes the number of tasks, the previous tasks (which will have been put into a publishing state) will be killed unnecessarily.. unnecessarily because the pendingCompletionTaskGroups data structure can handle different partition allocations but the actively reading taskGroups structure cannot.